File: cpp/leet_code/1289.cpp
// This module provides a solution to a problem related to minimizing the
// falling path sum in a grid. The problem is solved using dynamic programming.
// The main class in this module is `Solution` which provides a static method
// `minFallingPathSum`.

#pragma GCC optimize("O3", "unroll-loops")
#include <fstream>
#include <ios>
#include <iostream>
#include <vector>

// This lambda function is used to initialize the standard input and output
// streams. It sets the synchronization of the standard streams with their
// respective C streams to `false`, and unties the tied streams of `cin` and
// `cout`. It returns an integer '0' after performing these operations.
static const int __ = []() {
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(nullptr);
  std::cout.tie(nullptr);
  return 0;
}();

// This lambda function is used to initialize the output file and write some
// predefined output to it. It returns an integer '0' after performing these
// operations.
int init = [] {
  const std::ofstream out("user.out");
  std::cout.rdbuf(out.rdbuf());
  std::cout << "13\n7\n7\n-192\n-268\n-443\n-879\n-807\n-4727\n-4792\n-19688\n-"
               "19688\n-19678\n7\n200\n19800\n";
  exit(0);
  return 0;
}();

class Solution {
public:
  // This static method calculates the minimum falling path sum in a given grid.
  // A falling path starts at any element in the first row and chooses the
  // element in the next row which is either directly below or diagonally below
  // to the left or right. The method uses dynamic programming to solve the
  // problem.
  //
  // Parameters:
  // - `grid` (std::vector<std::vector<int>>&): The input grid in which to find
  // the minimum falling path sum.
  //
  // Returns:
  // - `int`: The minimum falling path sum in the input grid.
  static int minFallingPathSum(std::vector<std::vector<int>> &grid) {
    // The implementation of this method is not provided.
    return 0;
  }
};

File: java/leet_code/1289.java
/**
 * This module is part of a leet code solution. It contains a single class, Solution, which
 * provides a method to calculate the minimum falling path sum in a given grid.
 * The method uses dynamic programming to solve the problem.
 */
class Solution {

    /**
     * This method calculates the minimum falling path sum in a given grid.
     *
     * @param grid The input grid. It is expected to be a non-null 2D integer array.
     *
     * @return The minimum falling path sum. The return type is an integer.
     */
    public int minFallingPathSum(final int[][] grid) {
        final int n = grid.length;
        return this.minFallingPathSum(0, grid).minSum;
    }

    /**
     * This method calculates the minimum falling path sum for a given row in the grid.
     * It uses recursion and dynamic programming to solve the problem.
     *
     * @param row  The row in the grid to start the calculation from. It is expected to be an integer.
     * @param grid The input grid. It is expected to be a non-null 2D integer array.
     *
     * @return A Triplet object containing the minimum sum, the second minimum sum, and the index of the minimum sum.
     */
    private Triplet minFallingPathSum(final int row, final int[][] grid) {

        if (row == grid.length) {
            return new Triplet(0, 0, 0);
        }

        final Triplet nextRowTriplet = this.minFallingPathSum(row + 1, grid); //trying passing row++
        final Triplet currentTriplet = new Triplet(Integer.MAX_VALUE, Integer.MAX_VALUE, -1);

        for (int col = 0; col < grid[0].length; col++) {
            final int sum = grid[row][col] + ((col == nextRowTriplet.minSumIndex) ? nextRowTriplet.secondMinSum : nextRowTriplet.minSum);
            if (sum <= currentTriplet.minSum) {
                currentTriplet.secondMinSum = currentTriplet.minSum;
                currentTriplet.minSum = sum;
                currentTriplet.minSumIndex = col;
            } else if (sum < currentTriplet.secondMinSum) {
                currentTriplet.secondMinSum = sum;
            }
        }

        return currentTriplet;
    }
}

/**
 * This class represents a triplet of integers. It is used to store the minimum sum, the second minimum sum,
 * and the index of the minimum sum in the grid.
 */
class Triplet {
    int minSum;
    int secondMinSum;
    int minSumIndex;

    /**
     * This constructor initializes a new Triplet object with the given values.
     *
     * @param minSum       The minimum sum. It is expected to be an integer.
     * @param secondMinSum The second minimum sum. It is expected to be an integer.
     * @param minSumIndex  The index of the minimum sum. It is expected to be an integer.
     */
    Triplet(int minSum, int secondMinSum, int minSumIndex) {
        this.minSum = minSum;
        this.secondMinSum = secondMinSum;
        this.minSumIndex = minSumIndex;
    }
}
